File: mobile/app/src/main/java/ua/nure/cleaningservice/ui/profile/ServicesActivity.java
package ua.nure.cleaningservice.ui.profile;

import android.content.Intent;
import android.os.Bundle;
import android.util.Log;
import android.widget.ImageButton;

import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import java.util.ArrayList;
import java.util.List;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;
import ua.nure.cleaningservice.R;
import ua.nure.cleaningservice.data.Company;
import ua.nure.cleaningservice.data.Service;
import ua.nure.cleaningservice.network.JSONPlaceHolderApi;
import ua.nure.cleaningservice.network.NetworkService;
import ua.nure.cleaningservice.ui.add.AddServicesActivity;
import ua.nure.cleaningservice.ui.auth.MenuActivity;
import ua.nure.cleaningservice.ui.rva.ServicesRVA;

public class ServicesActivity extends AppCompatActivity {

    private static final String TAG = "ServicesActivity";

    private List<Service> mServices;
    private RecyclerView mRecyclerView;
    private JSONPlaceHolderApi mApi;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_services);

        mRecyclerView = (RecyclerView) findViewById(R.id.services_rv);

        ImageButton mBackButton = findViewById(R.id.back_btn);
        ImageButton mAddButton = findViewById(R.id.add_service_btn);

        mApi = NetworkService.getInstance().getApiService();

        LinearLayoutManager llm = new LinearLayoutManager(this);

        mRecyclerView.setLayoutManager(llm);
        mRecyclerView.setHasFixedSize(true);

        mBackButton.setOnClickListener((v) -> {
            navigateToScreen(MenuActivity.class);
            finish();
        });

        mAddButton.setOnClickListener((v) -> {
            navigateToScreen(AddServicesActivity.class);
            finish();
        });
    }

    @Override
    protected void onResume() {
        super.onResume();
        mServices = new ArrayList<>();
        initializeData();
    }

    private void initializeData() {
        String email = Company.getInstance().getEmail();
        String token = "Bearer " + Company.getInstance().getToken();
        mApi.getServiceData(token, email).enqueue(serviceCallback);
    }

    Callback<ArrayList<Service>> serviceCallback = new Callback<ArrayList<Service>>() {
        @Override
        public void onResponse(Call<ArrayList<Service>> call,
                Response<ArrayList<Service>> response) {
            if (!response.isSuccessful()) {
                System.out.println(response.code());
                return;
            }

            ArrayList<Service> serviceList = response.body();
            for (Service service : serviceList) {
                mServices.add(new Service(service.getId(), service.getName(),
                        service.getDescription(), service.getMinArea(),
                        service.getMaxArea(), service.getPlacementType(),
                        service.getPricePerMeter()));
            }
            initializeAdapter();
        }

        @Override
        public void onFailure(Call<ArrayList<Service>> call, Throwable t) {
            System.out.println(t);
            Log.i(TAG, t.getMessage());
        }
    };

    private void initializeAdapter() {
        ServicesRVA adapter = new ServicesRVA(this, mServices);
        mRecyclerView.setAdapter(adapter);
    }

    private void navigateToScreen(Class cls) {
        Intent intent = new Intent(ServicesActivity.this, cls);
        startActivity(intent);
    }
}

File: mobile/app/src/main/java/ua/nure/cleaningservice/ui/profile/ServicesActivity.kt
package ua.nure.cleaningservice.ui.profile

import android.content.Intent
import android.os.Build
import android.os.Bundle
import android.util.Log
import android.view.View
import android.widget.Button
import android.widget.EditText
import android.widget.ImageButton
import androidx.annotation.RequiresApi
import androidx.appcompat.app.AppCompatActivity
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response
import ua.nure.cleaningservice.R
import ua.nure.cleaningservice.data.User
import ua.nure.cleaningservice.data.Service
import ua.nure.cleaningservice.network.JsonPlaceHolderApi
import ua.nure.cleaningservice.network.NetworkService
import ua.nure.cleaningservice.ui.add.AddServicesActivity
import ua.nure.cleaningservice.ui.auth.MenuActivity
import ua.nure.cleaningservice.ui.edit.EditServicesActivity
import ua.nure.cleaningservice.ui.rva.PlacementsRVA
import ua.nure.cleaningservice.ui.rva.ServicesRVA
import ua.nure.cleaningservice.ui.util.AlertDialogUtil
import ua.nure.cleaningservice.ui.util.LoadingDialogUtil
import java.util.*

class ServicesActivity : AppCompatActivity() {

    private var mServices: MutableList<Service>? = null
    private var mRecyclerView: RecyclerView? = null
    private var mApi: JsonPlaceHolderApi? = null
    private lateinit var editButton: Button
    private lateinit var deleteButton: Button
    private val loadingDialog = LoadingDialogUtil(this@ServicesActivity)

    private var serviceId = -1

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_services)
        mRecyclerView = findViewById<View>(R.id.services_rv) as RecyclerView
        val mBackButton = findViewById<ImageButton>(R.id.back_btn)
        val mAddButton = findViewById<ImageButton>(R.id.add_service_btn)
        mApi = NetworkService.getInstance().apiService
        val llm = LinearLayoutManager(this)
        mRecyclerView!!.layoutManager = llm
        mRecyclerView!!.setHasFixedSize(true)

        mBackButton.setOnClickListener { v: View? ->
            navigateToScreen(MenuActivity::class.java)
            finish()
        }

        mAddButton.setOnClickListener { v: View? ->
            navigateToScreen(AddServicesActivity::class.java)
            finish()
        }

        editButton = findViewById(R.id.edit_service_btn)
        editButton.setOnClickListener {
            val input = EditText(this)
            val builder = AlertDialogUtil.getBaseAlertDialogBuilder(input, this@ServicesActivity)
            builder.setTitle(R.string.enter_service_number)
            builder.setPositiveButton("OK") { _, _ -> editService(Integer.parseInt(input.text.toString())) }
            builder.show()
        }

        deleteButton = findViewById(R.id.delete_service_btn)
        deleteButton.setOnClickListener {
            val input = EditText(this)
            val builder = AlertDialogUtil.getBaseAlertDialogBuilder(input, this@ServicesActivity)
            builder.setTitle(R.string.enter_service_number)
            builder.setPositiveButton("OK") { _, _ -> deleteService(Integer.parseInt(input.text.toString())) }
            builder.show()
        }
    }

    override fun onResume() {
        super.onResume()
        mServices = ArrayList()
        initializeData()
    }

    private fun initializeData() {
        loadingDialog.start()
        val email = User.getInstance().email
        val token = "Bearer " + User.getInstance().token
        mApi!!.getServiceData(token, email).enqueue(serviceCallback)
    }

    var serviceCallback: Callback<ArrayList<Service>> = object : Callback<ArrayList<Service>> {
        override fun onResponse(call: Call<ArrayList<Service>>,
                                response: Response<ArrayList<Service>>) {
            if (!response.isSuccessful) {
                println(response.code())
                loadingDialog.dismiss()
                return
            }
            val serviceList = response.body()!!
            for (service in serviceList) {
                mServices!!.add(Service(service.id, service.name,
                        service.description, service.minArea,
                        service.maxArea, service.placementType,
                        service.pricePerMeter))
            }
            initializeAdapter()
        }

        override fun onFailure(call: Call<ArrayList<Service>>, t: Throwable) {
            println(t)
            Log.i(TAG, t.message!!)
            loadingDialog.dismiss()
        }
    }

    private fun initializeAdapter() {
        val adapter = ServicesRVA(mServices)
        mRecyclerView!!.adapter = adapter
        loadingDialog.dismiss()
    }

    private fun navigateToScreen(cls: Class<*>) {
        val intent = Intent(this@ServicesActivity, cls)
        startActivity(intent)
    }

    private fun editService(id: Int) {
        val intent = Intent(this@ServicesActivity, EditServicesActivity::class.java)
        intent.putExtra("sId", id)
        startActivity(intent)
    }

    private fun deleteService(id: Int) {
        loadingDialog.start()
        serviceId = id
        mApi!!.deleteService("Bearer " + User.getInstance().token, id).enqueue(deleteCallback)
    }

    var deleteCallback: Callback<Service?> = object : Callback<Service?> {
        override fun onResponse(call: Call<Service?>, response: Response<Service?>) {
            println(response.body())
            loadingDialog.dismiss()
        }

        @RequiresApi(Build.VERSION_CODES.N)
        override fun onFailure(call: Call<Service?>, t: Throwable) {
            println(t)
            val adapter = mRecyclerView?.adapter as ServicesRVA
            adapter.mServices?.removeIf{ service -> service.id == serviceId }
            adapter.notifyDataSetChanged()
            loadingDialog.dismiss()
        }
    }

    companion object {
        private const val TAG = "ServicesActivity"
    }
}
